File: 102_binary_tree_level_order_traversal/levelOrder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    public:
        vector<vector<int>> levelOrder(TreeNode* root) {
            vector<vector<int>> res;
            bfs(root, 0, res);
            return res;
        }
    
    private:
        void bfs(TreeNode* root, int level, vector<vector<int>>& res) {
            if (root == 0) return;
    
            if (res.size() <= level) {
                res.push_back({});
            }
            res[level].push_back(root->val);
            bfs(root->left, level + 1, res);
            bfs(root->right, level + 1, res);
        }
    };

File: 102_binary_tree_level_order_traversal/levelOrder.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  const res = [];

  dfs(root, 0, res);
  return res;
};

function dfs(root, level, res) {
  if (root) {
    if (res.length <= level) {
      res.push([]);
    }

    res[level].push(root.val);
    dfs(root.left, level + 1, res);
    dfs(root.right, level + 1, res);
  }

  return;
}
